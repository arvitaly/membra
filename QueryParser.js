"use strict";
const graphql_1 = require("graphql");
const graphql_fields_info_1 = require("graphql-fields-info");
class QueryParser {
    constructor(schema) {
        this.schema = schema;
    }
    parse(literals, ...placeholders) {
        const query = literals.join("");
        const fieldsInfo = graphql_fields_info_1.fromQuery(query, this.schema);
        this.addIds(fieldsInfo.getFields());
        const text = "" + fieldsInfo.print();
        const fields = fieldsInfo.getFields();
        let type;
        if (fields[0].fields[0].isNode) {
            type = "node";
        }
        else if (fields[0].fields[0].isConnection) {
            type = "connection";
        }
        else {
            throw new Error("Unknown query type");
        }
        return {
            text,
            fields,
            type,
        };
    }
    addIds(fields) {
        fields.map((field) => {
            if (field.isNode) {
                if (!field.fields.find((f) => f.name === "id")) {
                    const node = {
                        name: { value: "id", kind: "Name" },
                        kind: "Field",
                    };
                    const sel = field.node.selectionSet;
                    if (sel) {
                        sel.selections.push(node);
                    }
                    field.fields.push({
                        args: [],
                        fields: [],
                        isFragment: false,
                        isConnection: false,
                        isNode: false,
                        node,
                        name: "id",
                        type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLID),
                    });
                }
            }
            this.addIds(field.fields);
        });
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = QueryParser;
