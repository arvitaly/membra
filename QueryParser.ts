import { FieldNode, GraphQLID, GraphQLNonNull, GraphQLSchema } from "graphql";
import { Fields, fromQuery, GraphQLFieldsInfo } from "graphql-fields-info";
import { IQuery } from "./typings";
class QueryParser {
    constructor(protected schema: GraphQLSchema) { }
    public parse<T>(literals: TemplateStringsArray, ...placeholders: any[]): IQuery<T> {
        const query: string = literals.join("");
        const fieldsInfo = fromQuery(query, this.schema);
        this.addIds(fieldsInfo.getFields());
        const text: string = "" + fieldsInfo.print();
        const fields = fieldsInfo.getFields();
        let type: "node" | "connection" | "nodeInterface";
        let nodeFields: Fields;
        if (fields[0].isInterface) {
            type = "nodeInterface";
            nodeFields = fieldsInfo.getNodeInterfaceFields();
        } else if (fields[0].fields[0].isNode) {
            type = "node";
            nodeFields = fieldsInfo.getQueryOneFields();
        } else if (fields[0].fields[0].isConnection) {
            type = "connection";
            nodeFields = fieldsInfo.getQueryConnectionFields();
        } else {
            throw new Error("Unknown query type");
        }
        return {
            text,
            fields,
            nodeFields,
            type,
        };
    }
    protected addIds(fields: Fields) {
        fields.map((field) => {
            if (field.isNode) {
                if (!field.fields.find((f) => f.name === "id")) {
                    const node: FieldNode = {
                        name: { value: "id", kind: "Name" },
                        kind: "Field",
                    };
                    const sel = (field.node as FieldNode).selectionSet;
                    if (sel) {
                        sel.selections.push(node);
                    }
                    field.fields.push({
                        args: [],
                        fields: [],
                        isFragment: false,
                        isConnection: false,
                        isNode: false,
                        node,
                        name: "id",
                        type: new GraphQLNonNull(GraphQLID),
                    });
                }
            }
            this.addIds(field.fields);
        });
    }
}
export default QueryParser;
