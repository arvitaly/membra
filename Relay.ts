import { Fields, fromQuery } from "graphql-fields-info";
import onemitter, { Onemitter } from "onemitter";
import { IQuery } from "./typings";
export interface IResolver {
    fetch(query: string, vars?: any, subscriptionId?: string): Promise<any>;
}
interface IData {
    id: string;
    value: any;
    ids: string[];
    query: IQuery;
    onemitter: Onemitter<any>;
}
interface ILiveQuery extends Onemitter<any> { }
class Relay {
    protected data: { [index: string]: IData } = {};
    protected id = 0;
    constructor(protected resolver: IResolver) { }
    public async live(query: IQuery, vars?: any): Promise<IData> {
        const id = this.getNewId();
        const data = await this.resolver.fetch(query.text, vars, id);
        const ids = this.getIds(data, query.fields);
        const o = onemitter();
        this.data[id] = {
            id,
            value: data,
            ids,
            query,
            onemitter: o,
        };

        setTimeout(() => {
            o.emit(data);
        });
        return this.data[id];
    }
    public addNode(dataId: string, globalId: string, value: any) {
        this.data[dataId].ids.push(globalId);
        const root = Object.keys(this.data[dataId].value)[0];
        const connection = Object.keys(this.data[dataId].value[root]).filter((o) => o !== "id")[0];
        this.data[dataId].value[root][connection].edges.push({
            node: value,
        });
        setTimeout(() => {
            this.data[dataId].onemitter.emit(this.data[dataId].value);
        });
    }
    public updateNode(dataId: string, globalId: string, value: any) {
        const rootField = this.data[dataId].query.fields[0].fields[0];
        const root = Object.keys(this.data[dataId].value)[0];
        const connection = Object.keys(this.data[dataId].value[root]).filter((o) => o !== "id")[0];
        const rootNode = this.data[dataId].value[root][connection];
        if (this.data[dataId].query.type === "node") {
            rootField.fields.map((field) => {
                if (typeof (value[field.name]) !== "undefined") {
                    rootNode[field.name] = value[field.name];
                }
            });
        } else {
            const edgesField = rootField.fields.find((f) => f.name === "edges");
            if (!edgesField) {
                throw new Error("Not found edges field in connection");
            }
            const nodeField = edgesField.fields.find((f) => f.name === "node");
            if (!nodeField) {
                throw new Error("Not found node field in connection");
            }
            rootNode.edges.filter((edge: any) => edge.node.id === globalId).map((edge: any) => {
                const node = edge.node;
                nodeField.fields.map((field) => {
                    if (typeof (value[field.name]) !== "undefined") {
                        node[field.name] = value[field.name];
                    }
                });
            });
        }
        setTimeout(() => {
            this.data[dataId].onemitter.emit(this.data[dataId].value);
        });
    }
    protected getIds(data: any, fields: Fields) {
        let ids: string[] = [];
        fields.map((field) => {
            if (field.isConnection) {
                ids = ids.concat(this.getIdsFromConnection(data[field.name], field.fields));
                return;
            }
            if (field.isNode) {
                ids.push(data[field.name].id);
            }
            ids = ids.concat(this.getIds(data[field.name], field.fields));
        });
        return ids;
    }
    protected getIdsFromConnection(data: any, fields: Fields) {
        let ids: string[] = [];
        const edgesField = fields.find((f) => f.name === "edges");
        if (!edgesField) {
            throw new Error("Not found edges field in connection");
        }
        const nodeField = edgesField.fields.find((f) => f.name === "node");
        if (!nodeField) {
            throw new Error("Not found node field in connection");
        }
        data.edges.map((edge: any) => {
            ids = ids.concat(this.getIds(edge.node, nodeField.fields));
        });
        return ids;
    }
    protected getNewId() {
        return "" + ++this.id;
    }
}
export default Relay;
