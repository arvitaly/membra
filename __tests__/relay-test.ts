import { toGlobalId } from "graphql-relay";
import schema from "./../__fixtures__/schema2";
import QueryParser from "./../QueryParser";
import Relay, { IResolver } from "./../Relay";

describe("Relay tests", () => {
    it("live", async () => {
        const parser = new QueryParser(schema);
        const query = parser.parse`            query Q1{
                viewer{
                    model1{
                        edges{
                            node{
                                field1
                                model2{
                                    field2
                                }
                            }
                        }
                    }
                }
            }`;
        const globalId1 = toGlobalId("Model2", "50");
        const resolver: IResolver = {
            fetch: jest.fn((q: string, vars?: any, subscriptionId?: string) => {
                expect({ q, vars, subscriptionId }).toMatchSnapshot();
                return {
                    viewer: {
                        id: toGlobalId("Viewer", ""),
                        model1: {
                            edges: [{
                                node: {
                                    id: globalId1,
                                    field1: "field1Value",
                                    model2: {
                                        id: toGlobalId("Model2", "100"),
                                        field2: 15,
                                    },
                                },
                            }],
                        },
                    },
                };
            }),
        };
        const globalId2 = toGlobalId("Model1", "533");
        const relay = new Relay(resolver);
        const data = await relay.live(query);
        let result = await data.onemitter.wait();
        expect(result).toMatchSnapshot();
        relay.addNode(data.id, globalId2, { field1: "field1Value2", excess1: "hi" });
        result = await data.onemitter.wait();
        expect(result).toMatchSnapshot();
        relay.updateNode(data.id, globalId1, { field1: "field1Value3", excess2: "hi2" });
        result = await data.onemitter.wait();
        expect(result).toMatchSnapshot();
        await relay.restoreAllLive();
        result = await data.onemitter.wait();
        expect(result).toMatchSnapshot();
    });
});
