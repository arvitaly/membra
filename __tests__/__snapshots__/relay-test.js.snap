exports[`Relay tests live 1`] = `
Object {
  "q": "query Q1 {
  viewer {
    model1 {
      edges {
        node {
          field1
          model2 {
            field2
            id
          }
          id
        }
      }
    }
    id
  }
}
",
  "subscriptionId": "1",
  "vars": undefined,
}
`;

exports[`Relay tests live 2`] = `
Object {
  "viewer": Object {
    "id": "Vmlld2VyOg==",
    "model1": Object {
      "edges": Array [
        Object {
          "node": Object {
            "field1": "field1Value",
            "id": "TW9kZWwyOjUw",
            "model2": Object {
              "field2": 15,
              "id": "TW9kZWwyOjEwMA==",
            },
          },
        },
      ],
    },
  },
}
`;

exports[`Relay tests live 3`] = `
Object {
  "viewer": Object {
    "id": "Vmlld2VyOg==",
    "model1": Object {
      "edges": Array [
        Object {
          "node": Object {
            "field1": "field1Value",
            "id": "TW9kZWwyOjUw",
            "model2": Object {
              "field2": 15,
              "id": "TW9kZWwyOjEwMA==",
            },
          },
        },
        Object {
          "node": Object {
            "field1": "field1Value2",
          },
        },
      ],
    },
  },
}
`;

exports[`Relay tests live 4`] = `
Object {
  "viewer": Object {
    "id": "Vmlld2VyOg==",
    "model1": Object {
      "edges": Array [
        Object {
          "node": Object {
            "field1": "field1Value3",
            "id": "TW9kZWwyOjUw",
            "model2": Object {
              "field2": 15,
              "id": "TW9kZWwyOjEwMA==",
            },
          },
        },
        Object {
          "node": Object {
            "field1": "field1Value2",
          },
        },
      ],
    },
  },
}
`;
